org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.xtend

```diff
@@ -43,73 +65,147 @@ class RequestManager {
 		executorService.shutdown()
 	}
 
-
 	/**
 	 * <p>
-	 * The given <i>write request</i> will be run first when <i>all running requests</i> completed.
+	 * Run the given asynchronous write request.
 	 * </p>
 	 * <p>
-	 * Currently <i>running requests</i> will be cancelled.
+	 * Write requests are executed exclusive.
 	 * </p>
 	 * <p>
-	 * A provided cancel indicator should implement {@link CancelIndicator} 
-	 * to let the given request to be cancelled by a write request.
+	 * An asynchronous write request:
+	 * <ul>
+	 * 	<li>can be canceled by any write request;</li>
+	 *  <li>cancels any asynchronous write request.</li>
+	 * </ul>
 	 * </p>
+	 * <p>
 	 */
 	def <V> CompletableFuture<V> runWrite((CancelIndicator)=>V writeRequest) {
-		semaphore.acquire(MAX_PERMITS)
-		try {
-			val result = writeRequest.apply([
-				return false
-			])
-			return CompletableFuture.completedFuture(result);
-		} catch (Throwable t) {
-			if (isCancelException(t)) {
-				LOGGER.info("request cancelled.")
-				throw new CancellationException()
-			}
-			throw t
-		} finally {
-			semaphore.release(MAX_PERMITS)
-		}
+		cancel
+		return runAsync(WRITE_PERMITS) [
```

Comment:
async locking for write request doesn't work, since it will jeopardize the order of changes.
They are coming in ordered, and we need to keep the. E.g. a read request coming directly after a write request, might be executed before it, running on outdated state.